tests/system/src/test/java/test/robot/javafx/scene/RobotTest.java

```diff
@@ -0,0 +1,342 @@
+    private static final int SIZE = 400;
+    private static final int WIDTH = 400;
+    private static final int HEIGHT = 300;
+
+    public static void main(String[] args) throws Exception {
+        RobotTest test = new RobotTest();
+        test.testKeyboard();
+        test.testMouseMove();
+        test.testMouseClick();
+        test.testMouseWheel();
+        test.testPixelCapture();
+        test.testScreenCapture();
+        exit();
+    }
+
+    @Test
+    public void testKeyboard() {
+        CountDownLatch onKeyTypedLatch = new CountDownLatch(1);
+        CountDownLatch setSceneLatch = new CountDownLatch(1);
+        TextField textField = new TextField();
+        InvalidationListener invalidationListener = observable -> setSceneLatch.countDown();
+        Platform.runLater(() -> {
+            textField.setOnKeyTyped(event -> onKeyTypedLatch.countDown());
+            scene = new Scene(new HBox(textField));
+            stage.sceneProperty().addListener(observable -> {
+                setSceneLatch.countDown();
+                stage.sceneProperty().removeListener(invalidationListener);
+            });
+            stage.setScene(scene);
+        });
+        waitForLatch(setSceneLatch, 5, "Timeout while waiting for scene to be set on stage.");
+        Platform.runLater(() -> {
+            int mouseX = (int) (scene.getWindow().getX() + scene.getX() +
+                    textField.getLayoutX() + textField.getLayoutBounds().getWidth() / 2);
+            int mouseY = (int) (scene.getWindow().getY() + scene.getY() +
+                    textField.getLayoutY() + textField.getLayoutBounds().getHeight() / 2);
+            robot.mouseMove(mouseX, mouseY);
+            robot.mouseClick(MouseButton.PRIMARY);
+            robot.keyPress(KeyCode.A);
+            robot.keyRelease(KeyCode.A);
+        });
+        waitForLatch(onKeyTypedLatch, 5, "Timeout while waiting for textField.onKeyTyped().");
+        Assert.assertEquals("letter 'a' should be typed by Robot", "a", textField.getText());
+    }
+
+    @Test
+    public void testMouseMove() {
+        CountDownLatch mouseMoveLatch = new CountDownLatch(1);
+        CountDownLatch setSceneLatch = new CountDownLatch(1);
+        InvalidationListener invalidationListener = observable -> setSceneLatch.countDown();
+        Platform.runLater(() -> {
+            scene = new Scene(new HBox(), SIZE, SIZE);
+            stage.sceneProperty().addListener(observable -> {
+                setSceneLatch.countDown();
+                stage.sceneProperty().removeListener(invalidationListener);
+            });
+            stage.setScene(scene);
+        });
+        waitForLatch(setSceneLatch, 5, "Timeout while waiting for scene to be set on stage.");
+        AtomicReference<Point2D> mousePosition = new AtomicReference<>();
+        Platform.runLater(() -> {
+            robot.mouseMove(50, 50);
+            mousePosition.set(robot.getMousePosition());
+            mouseMoveLatch.countDown();
```

Comment:
This is an example of a latch that will become unneeded if you use Util.runAndWait.